### PR TITLE
Tact pretty-printer

### LINKED ISSUE #162: Formatter
Currently, [tact-vscode](https://github.com/tact-lang/tact-vscode) does source code formatting using a plugin for [Prettier](https://prettier.io).
I believe a formatter should be a part of the language package. Also, we can have better control over how code is formatted.

Btw, since we now have a [Tree-sitter grammar for Tact](https://github.com/tact-lang/tree-sitter-tact), formatting can be done using [Topiary](https://topiary.tweag.io/).

The formatter should probably be aware of tactdoc comments: #232. Or at least keep the formatting of comments intact

### PR DESCRIPTION
**Pull request by heli0dus**
Towards #162 

- [ ] I have updated CHANGELOG.md
- [ ] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [ ] I have run all the tests locally and no test failure was reported
- [ ] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

**Comment by novusnota** [comment]
Nice progress!

One nit though: `package-lock.json` should be excluded. Please, use [Yarn v1](https://classic.yarnpkg.com/en/) for managing dependencies. It comes with newer versions of Node.js (16+ or even earlier), and uses `yarn.lock`, which we commit :)

File: src/formatter.ts [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+    }
+
+    ppASTStatementUntil(statement: ASTStatementUntil): string {
+        const condition = this.ppASTExpression(statement.condition);
+        const stmts = this.ppStatementBlock(statement.statements);
+        return `${this.indent()}do ${stmts} until (${condition});`;
+    }
+
+    ppASTStatementForEach(statement: ASTStatementForEach): string {
+        const header = `foreach (${statement.keyName}, ${statement.valueName} in ${statement.map.value})`
```
Comment (byakuren-hijiri) [inline src/formatter.ts]:
```suggestion
        const header = `foreach (${statement.keyName}, ${statement.valueName} in ${this.ppASTLvalueRef(statement.map)})`;
```

File: src/formatter.ts [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+            .map((stmt) => this.ppASTStatement(stmt))
+            .join("\n")
+        this.decreaseIndent()
+        const result = `{\n${stmntsFormatted}\n${this.indent()}}`;
+        return result;
+    }
+
+    ppASTStatementLet(statement: ASTStatementLet): string {
+        const expression = this.ppASTExpression(statement.expression);
+        return `${this.indent()}let ${statement.name}: ${this.ppASTTypeRef(statement.type)} = ${expression};`;
```
Comment (byakuren-hijiri) [inline src/formatter.ts]:
```suggestion
    const tyAnnotation = statement.type === null ? "" : `: ${this.ppASTTypeRef(statement.type)}`;
    return `${this.indent()}let ${statement.name}${tyAnnotation} = ${expression};`
```

File: src/formatter.ts [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+        const stmtsFormatted = func.statements
+            ? func.statements
+                  .map((stmt) => this.ppASTStatement(stmt))
+                  .join("\n")
+            : "";
+        const body = func.statements
+            ? `{\n${stmtsFormatted}`
+            : ";";
+        this.decreaseIndent();
+        return `${this.indent()}${attrsFormatted}fun ${func.name}(${argsFormatted})${returnType} ${body}\n${this.indent()}}`;
```
Comment (byakuren-hijiri) [inline src/formatter.ts]:
```suggestion
            const body = func.statements
      ? ` {\n${stmtsFormatted}\n${this.indent()}}`
      : ";";
    this.decreaseIndent();
    return `${this.indent()}${attrsFormatted}fun ${func.name}(${argsFormatted})${returnType}${body}`;
```

We also need a test for this, e.g.:
```
trait Q_e__dyxVu {
  abstract get fun z_(Uls9: String): String;
}
```

File: src/formatter.ts (hunk 2 above)
Comment (byakuren-hijiri) [reply src/formatter.ts]: We need a test for this like:
```
fun test(): Int {
    let a = 42;
    return a;
}
```

**Review by anton-trunov — APPROVED** [review]
Approving and merging this as an internal thing for now. Thanks @heli0dus!

**Pull request merged**